File: ruby/lib/hron/ast.rb
# frozen_string_literal: true

module Hron
  # Weekday enumeration (ISO 8601: Monday=1, Sunday=7)
  module Weekday
    MONDAY = :monday
    TUESDAY = :tuesday
    WEDNESDAY = :wednesday
    THURSDAY = :thursday
    FRIDAY = :friday
    SATURDAY = :saturday
    SUNDAY = :sunday

    ALL = [MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY, SUNDAY].freeze
    WEEKDAYS = [MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY].freeze
    WEEKEND = [SATURDAY, SUNDAY].freeze

    NUMBERS = {
      MONDAY => 1, TUESDAY => 2, WEDNESDAY => 3, THURSDAY => 4,
      FRIDAY => 5, SATURDAY => 6, SUNDAY => 7
    }.freeze

    CRON_DOW = {
      SUNDAY => 0, MONDAY => 1, TUESDAY => 2, WEDNESDAY => 3,
      THURSDAY => 4, FRIDAY => 5, SATURDAY => 6
    }.freeze

    NUMBER_TO_WEEKDAY = NUMBERS.invert.freeze

    PARSE_MAP = {
      "monday" => MONDAY, "mon" => MONDAY,
      "tuesday" => TUESDAY, "tue" => TUESDAY,
      "wednesday" => WEDNESDAY, "wed" => WEDNESDAY,
      "thursday" => THURSDAY, "thu" => THURSDAY,
      "friday" => FRIDAY, "fri" => FRIDAY,
      "saturday" => SATURDAY, "sat" => SATURDAY,
      "sunday" => SUNDAY, "sun" => SUNDAY
    }.freeze

    def self.number(day)
      NUMBERS[day]
    end

    def self.cron_dow(day)
      CRON_DOW[day]
    end

    def self.from_number(n)
      NUMBER_TO_WEEKDAY[n]
    end

    def self.try_parse(s)
      PARSE_MAP[s.downcase]
    end

    def self.to_s(day)
      day.to_s
    end
  end

  # Month name enumeration
  module MonthName
    JAN = :jan
    FEB = :feb
    MAR = :mar
    APR = :apr
    MAY = :may
    JUN = :jun
    JUL = :jul
    AUG = :aug
    SEP = :sep
    OCT = :oct
    NOV = :nov
    DEC = :dec

    ALL = [JAN, FEB, MAR, APR, MAY, JUN, JUL, AUG, SEP, OCT, NOV, DEC].freeze

    NUMBERS = {
      JAN => 1, FEB => 2, MAR => 3, APR => 4, MAY => 5, JUN => 6,
      JUL => 7, AUG => 8, SEP => 9, OCT => 10, NOV => 11, DEC => 12
    }.freeze

    NUMBER_TO_MONTH = NUMBERS.invert.freeze

    PARSE_MAP = {
      "january" => JAN, "jan" => JAN,
      "february" => FEB, "feb" => FEB,
      "march" => MAR, "mar" => MAR,
      "april" => APR, "apr" => APR,
      "may" => MAY,
      "june" => JUN, "jun" => JUN,
      "july" => JUL, "jul" => JUL,
      "august" => AUG, "aug" => AUG,
      "september" => SEP, "sep" => SEP,
      "october" => OCT, "oct" => OCT,
      "november" => NOV, "nov" => NOV,
      "december" => DEC, "dec" => DEC
    }.freeze

    def self.number(month)
      NUMBERS[month]
    end

    def self.from_number(n)
      NUMBER_TO_MONTH[n]
    end

    def self.try_parse(s)
      PARSE_MAP[s.downcase]
    end

    def self.to_s(month)
      month.to_s
    end
  end

  # Interval unit (minutes or hours)
  module IntervalUnit
    MIN = :min
    HOURS = :hours

    def self.to_s(unit)
      unit.to_s
    end
  end

  # Ordinal position (first, second, etc.)
  module OrdinalPosition
    FIRST = :first
    SECOND = :second
    THIRD = :third
    FOURTH = :fourth
    FIFTH = :fifth
    LAST = :last

    TO_N = {
      FIRST => 1, SECOND => 2, THIRD => 3, FOURTH => 4, FIFTH => 5
    }.freeze

    def self.to_n(ord)
      TO_N[ord]
    end

    def self.to_s(ord)
      ord.to_s
    end
  end

  # Time of day (hour and minute)
  TimeOfDay = Data.define(:hour, :minute) do
    def to_s
      format("%02d:%02d", hour, minute)
    end
  end

  # --- Day filter variants ---

  DayFilterEvery = Data.define
  DayFilterWeekday = Data.define
  DayFilterWeekend = Data.define
  DayFilterDays = Data.define(:days) # days: Array<Weekday>

  # --- Day of month spec ---

  SingleDay = Data.define(:day)
  DayRange = Data.define(:start, :end_day) # end_day to avoid Ruby keyword

  # --- Month target variants ---

  DaysTarget = Data.define(:specs) # specs: Array<DayOfMonthSpec>
  LastDayTarget = Data.define
  LastWeekdayTarget = Data.define

  # --- Year target variants ---

  YearDateTarget = Data.define(:month, :day)
  YearOrdinalWeekdayTarget = Data.define(:ordinal, :weekday, :month)
  YearDayOfMonthTarget = Data.define(:day, :month)
  YearLastWeekdayTarget = Data.define(:month)

  # --- Date spec variants ---

  NamedDate = Data.define(:month, :day)
  IsoDate = Data.define(:date) # date: String (YYYY-MM-DD)

  # --- Exception spec variants ---

  NamedException = Data.define(:month, :day)
  IsoException = Data.define(:date)

  # --- Until spec variants ---

  IsoUntil = Data.define(:date)
  NamedUntil = Data.define(:month, :day)

  # --- Schedule expression variants ---

  IntervalRepeat = Data.define(:interval, :unit, :from_time, :to_time, :day_filter)
  DayRepeat = Data.define(:interval, :days, :times)
  WeekRepeat = Data.define(:interval, :days, :times)
  MonthRepeat = Data.define(:interval, :target, :times)
  OrdinalRepeat = Data.define(:interval, :ordinal, :day, :times)
  SingleDateExpr = Data.define(:date, :times)
  YearRepeat = Data.define(:interval, :target, :times)

  # --- Schedule data (top-level) ---

  ScheduleData = Data.define(:expr, :timezone, :except, :until, :anchor, :during) do
    def initialize(expr:, timezone: nil, except: [], until: nil, anchor: nil, during: [])
      super
    end
  end

  # --- Helper functions ---

  def self.expand_day_spec(spec)
    case spec
    when SingleDay
      [spec.day]
    when DayRange
      (spec.start..spec.end_day).to_a
    else
      []
    end
  end

  def self.expand_month_target(target)
    case target
    when DaysTarget
      target.specs.flat_map { |spec| expand_day_spec(spec) }
    else
      []
    end
  end
end

File: python/src/hron/_ast.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class Weekday(Enum):
    MONDAY = "monday"
    TUESDAY = "tuesday"
    WEDNESDAY = "wednesday"
    THURSDAY = "thursday"
    FRIDAY = "friday"
    SATURDAY = "saturday"
    SUNDAY = "sunday"

    @property
    def number(self) -> int:
        """ISO 8601 day number: Monday=1, Sunday=7."""
        return _WEEKDAY_NUMBERS[self]

    @property
    def cron_dow(self) -> int:
        """Cron DOW number: Sunday=0, Monday=1, ..., Saturday=6."""
        return _CRON_DOW[self]

    @classmethod
    def from_number(cls, n: int) -> Weekday | None:
        return _NUMBER_TO_WEEKDAY.get(n)

    @classmethod
    def try_parse(cls, s: str) -> Weekday | None:
        return _WEEKDAY_PARSE.get(s.lower())

    def __str__(self) -> str:
        return self.value


_WEEKDAY_NUMBERS = {
    Weekday.MONDAY: 1,
    Weekday.TUESDAY: 2,
    Weekday.WEDNESDAY: 3,
    Weekday.THURSDAY: 4,
    Weekday.FRIDAY: 5,
    Weekday.SATURDAY: 6,
    Weekday.SUNDAY: 7,
}

_CRON_DOW = {
    Weekday.SUNDAY: 0,
    Weekday.MONDAY: 1,
    Weekday.TUESDAY: 2,
    Weekday.WEDNESDAY: 3,
    Weekday.THURSDAY: 4,
    Weekday.FRIDAY: 5,
    Weekday.SATURDAY: 6,
}

_NUMBER_TO_WEEKDAY = {v: k for k, v in _WEEKDAY_NUMBERS.items()}

_WEEKDAY_PARSE: dict[str, Weekday] = {
    "monday": Weekday.MONDAY,
    "mon": Weekday.MONDAY,
    "tuesday": Weekday.TUESDAY,
    "tue": Weekday.TUESDAY,
    "wednesday": Weekday.WEDNESDAY,
    "wed": Weekday.WEDNESDAY,
    "thursday": Weekday.THURSDAY,
    "thu": Weekday.THURSDAY,
    "friday": Weekday.FRIDAY,
    "fri": Weekday.FRIDAY,
    "saturday": Weekday.SATURDAY,
    "sat": Weekday.SATURDAY,
    "sunday": Weekday.SUNDAY,
    "sun": Weekday.SUNDAY,
}


class MonthName(Enum):
    JAN = "jan"
    FEB = "feb"
    MAR = "mar"
    APR = "apr"
    MAY = "may"
    JUN = "jun"
    JUL = "jul"
    AUG = "aug"
    SEP = "sep"
    OCT = "oct"
    NOV = "nov"
    DEC = "dec"

    @property
    def number(self) -> int:
        return _MONTH_NUMBERS[self]

    @classmethod
    def try_parse(cls, s: str) -> MonthName | None:
        return _MONTH_PARSE.get(s.lower())

    def __str__(self) -> str:
        return self.value


_MONTH_NUMBERS = {
    MonthName.JAN: 1,
    MonthName.FEB: 2,
    MonthName.MAR: 3,
    MonthName.APR: 4,
    MonthName.MAY: 5,
    MonthName.JUN: 6,
    MonthName.JUL: 7,
    MonthName.AUG: 8,
    MonthName.SEP: 9,
    MonthName.OCT: 10,
    MonthName.NOV: 11,
    MonthName.DEC: 12,
}

_MONTH_PARSE: dict[str, MonthName] = {
    "january": MonthName.JAN,
    "jan": MonthName.JAN,
    "february": MonthName.FEB,
    "feb": MonthName.FEB,
    "march": MonthName.MAR,
    "mar": MonthName.MAR,
    "april": MonthName.APR,
    "apr": MonthName.APR,
    "may": MonthName.MAY,
    "june": MonthName.JUN,
    "jun": MonthName.JUN,
    "july": MonthName.JUL,
    "jul": MonthName.JUL,
    "august": MonthName.AUG,
    "aug": MonthName.AUG,
    "september": MonthName.SEP,
    "sep": MonthName.SEP,
    "october": MonthName.OCT,
    "oct": MonthName.OCT,
    "november": MonthName.NOV,
    "nov": MonthName.NOV,
    "december": MonthName.DEC,
    "dec": MonthName.DEC,
}


class IntervalUnit(Enum):
    MIN = "min"
    HOURS = "hours"

    def __str__(self) -> str:
        return self.value


class OrdinalPosition(Enum):
    FIRST = "first"
    SECOND = "second"
    THIRD = "third"
    FOURTH = "fourth"
    FIFTH = "fifth"
    LAST = "last"

    def to_n(self) -> int:
        return _ORDINAL_TO_N[self]

    def __str__(self) -> str:
        return self.value


_ORDINAL_TO_N: dict[OrdinalPosition, int] = {
    OrdinalPosition.FIRST: 1,
    OrdinalPosition.SECOND: 2,
    OrdinalPosition.THIRD: 3,
    OrdinalPosition.FOURTH: 4,
    OrdinalPosition.FIFTH: 5,
}


class NearestDirection(Enum):
    """Direction for nearest weekday (hron extension beyond cron W)."""

    NEXT = "next"  # Always prefer following weekday (can cross to next month)
    PREVIOUS = "previous"  # Always prefer preceding weekday (can cross to prev month)

    def __str__(self) -> str:
        return self.value


@dataclass(frozen=True, slots=True)
class TimeOfDay:
    hour: int
    minute: int

    def __str__(self) -> str:
        return f"{self.hour:02d}:{self.minute:02d}"


# --- Day filter ---


@dataclass(frozen=True, slots=True)
class DayFilterEvery:
    pass


@dataclass(frozen=True, slots=True)
class DayFilterWeekday:
    pass


@dataclass(frozen=True, slots=True)
class DayFilterWeekend:
    pass


@dataclass(frozen=True, slots=True)
class DayFilterDays:
    days: tuple[Weekday, ...]


DayFilter = DayFilterEvery | DayFilterWeekday | DayFilterWeekend | DayFilterDays


# --- Day of month spec ---


@dataclass(frozen=True, slots=True)
class SingleDay:
    day: int


@dataclass(frozen=True, slots=True)
class DayRange:
    start: int
    end: int


DayOfMonthSpec = SingleDay | DayRange


# --- Month target ---


@dataclass(frozen=True, slots=True)
class DaysTarget:
    specs: tuple[DayOfMonthSpec, ...]


@dataclass(frozen=True, slots=True)
class LastDayTarget:
    pass


@dataclass(frozen=True, slots=True)
class LastWeekdayTarget:
    pass


@dataclass(frozen=True, slots=True)
class NearestWeekdayTarget:
    """Nearest weekday to a given day of month.

    Standard (direction=None): never crosses month boundary (cron W compatibility).
    Directional (direction=Some): can cross month boundary.
    """

    day: int
    direction: NearestDirection | None = None


MonthTarget = DaysTarget | LastDayTarget | LastWeekdayTarget | NearestWeekdayTarget


# --- Year target ---


@dataclass(frozen=True, slots=True)
class YearDateTarget:
    month: MonthName
    day: int


@dataclass(frozen=True, slots=True)
class YearOrdinalWeekdayTarget:
    ordinal: OrdinalPosition
    weekday: Weekday
    month: MonthName


@dataclass(frozen=True, slots=True)
class YearDayOfMonthTarget:
    day: int
    month: MonthName


@dataclass(frozen=True, slots=True)
class YearLastWeekdayTarget:
    month: MonthName


YearTarget = (
    YearDateTarget | YearOrdinalWeekdayTarget | YearDayOfMonthTarget | YearLastWeekdayTarget
)


# --- Date spec ---


@dataclass(frozen=True, slots=True)
class NamedDate:
    month: MonthName
    day: int


@dataclass(frozen=True, slots=True)
class IsoDate:
    date: str


DateSpec = NamedDate | IsoDate


# --- Exception ---


@dataclass(frozen=True, slots=True)
class NamedException:
    month: MonthName
    day: int


@dataclass(frozen=True, slots=True)
class IsoException:
    date: str


ExceptionSpec = NamedException | IsoException


# --- Until spec ---


@dataclass(frozen=True, slots=True)
class IsoUntil:
    date: str


@dataclass(frozen=True, slots=True)
class NamedUntil:
    month: MonthName
    day: int


UntilSpec = IsoUntil | NamedUntil


# --- Schedule expressions ---


@dataclass(frozen=True, slots=True)
class IntervalRepeat:
    interval: int
    unit: IntervalUnit
    from_time: TimeOfDay
    to_time: TimeOfDay
    day_filter: DayFilter | None


@dataclass(frozen=True, slots=True)
class DayRepeat:
    interval: int
    days: DayFilter
    times: tuple[TimeOfDay, ...]


@dataclass(frozen=True, slots=True)
class WeekRepeat:
    interval: int
    days: tuple[Weekday, ...]
    times: tuple[TimeOfDay, ...]


@dataclass(frozen=True, slots=True)
class MonthRepeat:
    interval: int
    target: MonthTarget
    times: tuple[TimeOfDay, ...]


@dataclass(frozen=True, slots=True)
class OrdinalRepeat:
    interval: int
    ordinal: OrdinalPosition
    day: Weekday
    times: tuple[TimeOfDay, ...]


@dataclass(frozen=True, slots=True)
class SingleDateExpr:
    date: DateSpec
    times: tuple[TimeOfDay, ...]


@dataclass(frozen=True, slots=True)
class YearRepeat:
    interval: int
    target: YearTarget
    times: tuple[TimeOfDay, ...]


ScheduleExpr = (
    IntervalRepeat
    | DayRepeat
    | WeekRepeat
    | MonthRepeat
    | OrdinalRepeat
    | SingleDateExpr
    | YearRepeat
)


# --- Schedule data (top-level) ---


@dataclass(slots=True)
class ScheduleData:
    expr: ScheduleExpr
    timezone: str | None = None
    except_: tuple[ExceptionSpec, ...] = ()
    until: UntilSpec | None = None
    anchor: str | None = None
    during: tuple[MonthName, ...] = ()


def new_schedule_data(expr: ScheduleExpr) -> ScheduleData:
    return ScheduleData(expr=expr)


# --- Helper functions ---


ALL_WEEKDAYS: tuple[Weekday, ...] = (
    Weekday.MONDAY,
    Weekday.TUESDAY,
    Weekday.WEDNESDAY,
    Weekday.THURSDAY,
    Weekday.FRIDAY,
)

ALL_WEEKEND: tuple[Weekday, ...] = (Weekday.SATURDAY, Weekday.SUNDAY)


def expand_day_spec(spec: DayOfMonthSpec) -> list[int]:
    match spec:
        case SingleDay(day=d):
            return [d]
        case DayRange(start=s, end=e):
            return list(range(s, e + 1))


def expand_month_target(target: MonthTarget) -> list[int]:
    match target:
        case DaysTarget(specs=specs):
            result: list[int] = []
            for spec in specs:
                result.extend(expand_day_spec(spec))
            return result
        case _:
            return []
